joining/src/main/scala/akka/cluster/bootstrap/dns/HeadlessServiceDnsBootstrap.scala

```diff
@@ -0,0 +1,278 @@
+      // TODO would be nice if we can discover this via DNS as well? and leave option to override here
+      val targetPort = settings.httpContactPointPort
+      val baseUri = Uri("http", Uri.Authority(Uri.Host(contactPoint), targetPort))
+      ensureProbing(baseUri)
+    }
+  }
+
+  private def onNoSeedNodesObtainedWithinStableDeadline(contactPoint: Uri): Unit = {
+    val dnsRecordsAreStable = lastContactsObservation.isPastStableMargin(settings, timeNow())
+    if (dnsRecordsAreStable) {
+      lastContactsObservation.selfAddressIfAbleToJoinItself(context.system) match {
+        case Some(allowedToJoinSelfAddress) ⇒
+          log.info(
+              "Initiating new cluster, self-joining [{}], as this node has the LOWEST address out of: [{}]! " +
+              "Other nodes are expected to locate this cluster via continued contact-point probing.")
+
+          cluster.join(allowedToJoinSelfAddress)
+
+          context.stop(self) // the bootstraping is complete
+        case None ⇒
+          log.info(
+              "Exceeded stable margins without locating seed-nodes, however this node is NOT the lowest address out " +
+              "of the discovered IPs in this deployment, thus NOT joining self. Expecting node ({}) to perform the self-join " +
+              "and initiate the cluster.", lastContactsObservation.lowestAddressContactPoint)
+
+        // nothing to do anymore, the probing will continue until the lowest addressed node decides to join itself.
+        // note, that due to DNS changes this may still become this node! We'll then await until the dns stableMargin
+        // is exceeded and would decide to try joining self again (same code-path), that time successfully though.
+      }
+
+    } else {
+      // TODO throttle this logging? It may be caused by any of the probing actors
+      log.info(
+          "DNS observation has changed more recently than the dnsStableMargin({}) allows (at: {}), not considering to join myself. " +
+          "This process will be retried.", settings.dnsStableMargin, new Date(lastContactsObservation.observedAt))
+    }
+  }
+
+  private def ensureProbing(baseUri: Uri): ActorRef = {
+    val childActorName = s"contactPointProbe-${baseUri.authority.host}-${baseUri.authority.port}"
+    log.info("Ensuring probing actor: " + childActorName)
+    context.child(childActorName) match {
+      case Some(contactPointProbingChild) ⇒
+        contactPointProbingChild
+      case None ⇒
+        val props = HttpContactPointBootstrap.props(settings, self, baseUri)
+        context.actorOf(props, childActorName)
+    }
+  }
+
+  private def scheduleNextResolve(serviceName: String, resolveTimeout: FiniteDuration): Unit =
+    timers.startSingleTimer(TimerKeyResolveDNS, Internal.AttemptResolve(serviceName, resolveTimeout),
+      settings.dnsResolveInterval)
+
+  private def resolve(name: String, resolveTimeout: FiniteDuration): Future[DnsResolvedTo] = {
```

Comment:
Would it be possible to make this modular? The real concept here is just service discovery, right? It would be nice if we could provide our own module that actually ties this to DNS SRV or whatever else we'd like to.  As an example, I mean something like:

```
trait ServiceDiscovery {
  def lookup(name: String, resolveTimeout: Future[ResolvedTo] 
}

// then perhaps you provide a default impl with the library

class DnsSrvServiceDiscovery {
  def lookup(name: String, resolveTimeout: Future[ResolvedTo] =
    dns.ask(Dns.Resolve(name))(resolveTimeout) map {
      ...
    }
}

// and in reference.conf

akka.management.service-discovery-class = "akka.cluster.bootstrap.dns.DnsSrvServiceDiscovery"

```

Felix and I could then provide our own implementation in `reactive-lib` that inspects its runtime environment to decide if discovery is done via DNS SRV, HTTP, or whatever else.

Reply:
I think that is a good idea.

Impl can optionally take ActorSystem as param, as we do in many other places.

Reply:
Btw, that will probably also simplify testing, using a stub.

Reply:
Yeah I’ll do that :)